pkg/server/list.go

```diff
@@ -57,6 +61,10 @@ func (l *LimitedServer) list(ctx context.Context, r *etcdserverpb.RangeRequest)
 		rev, resp.Count, err = l.backend.Count(ctx, prefix, start, revision)
 		logrus.Tracef("LIST COUNT key=%s, end=%s, revision=%d, currentRev=%d count=%d", r.Key, r.RangeEnd, revision, rev, resp.Count)
 		resp.Header = txnHeader(rev)
+
+		if err == nil && resp.Count > maxItems && (r.Limit == 0 || r.Limit > maxItems) && prefix == start {
+			return nil, errors.New("dataset size exceeds the allowed limit, please use limit or rage end to reduce the number of items")
+		}
```

Comment:
etcd doesn't enforce this limitation as far as I know, so in general I would be against doing something that etcd doesn't. If the apiserver does this for us on the Kubernetes side, why should we do it here?

Reply:
@brandond because users should forget it, and a wrong query should kill the cluster. I can totally agree with your point, i killed my cluster a few times because i forgot filtering or limiting my query :D